src/app/Dashboard/Charts/ChartController.tsx

```diff
@@ -0,0 +1,190 @@
+ */
+
+import { ApiService } from '@app/Shared/Services/Api.service';
+import { NotificationCategory, NotificationChannel } from '@app/Shared/Services/NotificationChannel.service';
+import { SettingsService } from '@app/Shared/Services/Settings.service';
+import { NO_TARGET, Target, TargetService } from '@app/Shared/Services/Target.service';
+import {
+  BehaviorSubject,
+  catchError,
+  concatMap,
+  finalize,
+  first,
+  map,
+  merge,
+  Observable,
+  of,
+  pairwise,
+  ReplaySubject,
+  Subscription,
+  switchMap,
+  throttleTime,
+} from 'rxjs';
+
+export const RECORDING_NAME = 'dashboard_metrics';
+
+export class ChartController {
+  private readonly _refCount$ = new BehaviorSubject<number>(0);
+  private readonly _updates$ = new ReplaySubject<void>(1);
+  private readonly _hasRecording$ = new ReplaySubject<boolean>(1);
+  private readonly _subscriptions: Subscription[] = [];
+
+  constructor(
+    private readonly _api: ApiService,
+    private readonly _target: TargetService,
+    private readonly _notifications: NotificationChannel,
+    private readonly _settings: SettingsService
+  ) {
+    this._subscriptions.push(
+      this._refCount$
+        .pipe(
+          map((v) => v > 0),
+          pairwise()
+        )
+        .subscribe((v) => {
+          const prev = v[0];
+          const curr = v[1];
+          if (prev && !curr) {
+            // last subscriber left
+            this._tearDown();
+          }
+          if (!prev && curr) {
+            // first subscriber joined
+            this._init();
+          }
+        })
+    );
+  }
+
+  hasActiveRecording(): Observable<boolean> {
+    return this._hasRecording$.asObservable();
+  }
+
+  attach(): Observable<number> {
+    this._refCount$.next(this._refCount$.value + 1);
+    return this._refCount$.asObservable().pipe(finalize(() => this._refCount$.next(this._refCount$.value - 1)));
+  }
+
+  requestRefresh(): void {
+    this._updates$.next();
+  }
+
+  private _hasRecording(target: Target): Observable<boolean> {
+    if (target === NO_TARGET) {
+      return of(false);
+    }
+
+    return this._api
+      .graphql<CountResponse>(
+        `
+          query ActiveRecordingsForAutomatedAnalysis($connectUrl: String) {
+            targetNodes(filter: { name: $connectUrl }) {
+              recordings {
+                active (filter: {
+                  labels: ["origin=${RECORDING_NAME}"],
+                  state: "RUNNING",
+                }) {
+                aggregate {
+                  count
+                }
+              }
+            }
+            }
+          }`,
+        { connectUrl: target.connectUrl }
+      )
+      .pipe(
+        map((resp) => {
+          const nodes = resp?.data?.targetNodes;
+          if (nodes.length === 0) {
+            return false;
+          }
+          const count = nodes[0]?.recordings?.active?.aggregate?.count;
+          return count > 0;
+        }),
+        catchError((_) => of(false)),
+        first()
```

Comment:
I think we can remove `first` here as `context.api.graphql` internally calls it already?